### ISSUE TITLE
Sometimes users appear as Anonymous when they should not

### ISSUE THREAD

**Issue by barbaldo** [liveblocks/liveblocks#2045]
We've observed that sometimes users in our product appear as anonymous when they should have their name. The situation is as follows:
In this room there are two users: Marco and Brian, in the comment section Marco has tagged Brian a few times. At this point, it's my understanding (which may be incorrent) that LiveBlocks will trigger the users discovery endpoint to try and figure out who these people are, and that's what happens but there seems to be a mistake.

<img width="1179" alt="Screenshot 2024-11-07 at 16 55 01" src="https://github.com/user-attachments/assets/81753fdc-2603-44d3-a591-010033143a03">

On the left side is a screenshot of the application, showing Anonymous (Marco) tagging Brian (who is correctly shown). On the right side is an excerpt from the network tab with the http call triggered by resolveUsers, the issue seems to be that only Brian's user_id is being passed and somehow Marco's gets lost.

**Comment by barbaldo** [comment]
Addendum: this situation persists until the page is reloaded, after that everything seems to be fine for a few days, the the problem shows up again.

**Comment by nvie** [comment]
This does indeed look wrong! Thanks for filing this issue.

Just to confirm, are you on the latest version of our packages (2.11.0)? To upgrade, you can run `npx create-liveblocks-app@latest --upgrade`, which will bump all Liveblocks dependencies in your project to the same, latest, version.

I will take this back to the team if they have an idea of what might be up here.

**Comment by marcbouchenoire** [comment]
Hey @barbaldo, thanks for the detailed report!

You're correct that when user IDs become visible on screen, we will trigger your `resolveUsers` callback with them. But we only provide `resolveUsers` with **unknown** user IDs, meaning that if they were previously resolved (if they were visible in another part of the app previously for example), we won't pass them to your `resolveUsers` again.

Your issue could be due to multiple reasons so I have a few questions to help us debug this further:
- Is the screenshot a scenario where you've just arrived on this screen and saw these two user IDs both for the first time?
- Could you share your `resolveUsers` callback and if your endpoint triggered by it does anything with the provided user IDs?

Thanks!

**Comment by barbaldo** [comment]
Thanks for the quick replies! I'll answer to both in order:
@nvie I have not, I was waiting for this to be resolved: https://github.com/liveblocks/liveblocks/issues/2010 but I'll give it a go
@marcbouchenoire Not necessarily, the missing user Marco could have been displayed earlier in the app. The second user for sure just appeared.
I can share some of it! (Ignore the ugly code and naming conventions, we're rushing)

```
resolveUsers={async ({ userIds }) => {
          const res = await apigwFetch("lb/users", {
            method: "POST",
            body: {
              company_id: current_company.id,
              framework_id: current_framework.id,
              userIds: userIds,
            },
          });
          let actual_users = userIds.map((user: any) => {
            let u = res.find((u: any) => u.id === user);
            return {
              name: `${u.first_name} ${u.last_name}`,
              avatar: u.avatar,
            };
          });
          return [...actual_users];
        }}
```

Basically what happens is: whenever the method is called, we call and endpoint made specifically for that where our DB gets called, users are fetched based on the IDs passed, then we return an object with name and avatar. Nothing too crazy but hope it helps!

**Comment by marcbouchenoire** [comment]
One possibility could be that somehow one part of this is erroring (either in `resolveUsers` directly or from your endpoint), and currently if `resolveUsers` errors/rejects (or returns `undefined` in its array for a specific ID, but that can't happen with your code), we mark the user IDs as anonymous and won't retry them later (until a page reload). The screenshot scenario would be that Marco was visible earlier but resolving it failed, then arriving on this screen only triggered a resolve for the second one because Marco was already marked as known but anonymous.

If you can easily repro the issue, would it be possible to add some logging to see if that's the case?

**Comment by barbaldo** [comment]
I'll ad some logging to see if I get some additional info, unfortunately the issue isn't easy to reproduce. It seems to show up a little bit randomly after some time but I'll keep an eye on it.

**Comment by barbaldo** [comment]
@marcbouchenoire after a period of increased logging, I'm able to confirm with some degree of certainty that the issue still happens and it looks like it's not our fault. Meaning that as far as we can tell from the logs, every time the LiveblocksProvider calls our endpoint requesting a specific number of user IDs, the endpoint always sends back the correct data. Yet we're still getting reports of people appearing as Anonymous.

Do you have any other suggestion on something to do or try to figure it out?

**Comment by marcbouchenoire** [comment]
@barbaldo I'm not sure unfortunately, the data does appears correct on your side but on ours we haven't seen regressions from our tests or had any similar reports. I still think the issue could be on our side with a niche sequence of events that isn't tested (even though it doesn't seem the case from your previous comments). The best next step would be to find a simple repro setup that we can use and learn from, if you think that's possible?

**Comment by barbaldo** [comment]
@marcbouchenoire sorry I haven't been replying ( much work to do ). I'm not sure where to even start, it happens incredibly randomly on our end too so I'll try to gather more data on this.

**Comment by barbaldo** [comment]
@marcbouchenoire the only thing we've managed to figure out so far is that our customers tend to keep really long sessions, as in our auth allows to stay logged up to 15 days and it seems that users keep the same browser tab open for that long. So IDK maybe the LiveBlocks cache starts crashing after that long? Seems unlikely but I'm spitballing

**Comment by marcbouchenoire** [comment]
@barbaldo The cache is just in-memory and it has not time-based logic so I would be confident to exclude it from the issue, but otherwise sorry I can't help much either, without a repro I haven't been able to find ideas for possible culprits.

**Comment by barbaldo** [comment]
@marcbouchenoire I was able to prevent the issue from happening, doesn't mean it's solved though. But at least our users stopped pestering me about it :) . I may try to do a repro in the future, in the meantime thanks for all the help.

**Comment by barbaldo** [comment]
@onx2 not really a fix more like a hacky solution. Basically we force a page reload everytime a user has been inactive for some time. It's a terrible solution but so far has prevented the issue from happening

**Comment by marcbouchenoire** [comment]
@onx2 Are you also not able to reproduce the issue consistently? I'd love to isolate the problem and bring it into our tests to understand it.

**Comment by marcbouchenoire** [comment]
Thanks 🙏

> Also a user pointed out that refreshing the browser didn't clear the issue for them.

Our caching is only in memory so it won't be kept across reloads and `resolveUsers` will be called again after one, so it would be a sign that the issue happens within your `resolveUsers` logic and how it doesn't return user info for one/some ID(s). I'd recommend adding ways to see exactly what you return from any `resolveUsers` call in case you repro the issue randomly.

**Comment by onx2** [comment]
Good to know! So far my most reasonable ideas for this problem, generally speaking are:
- the scope that user was in no longer exists or they no longer are a part of it
- the user is an admin with rights to comment but doesn’t show up in the scoped users results
- the user is an external contributor that can comment but isn’t a part of the scoped results
- the liveblockUsers result set isn’t returning all of the users in the scope and isn’t paginating so some are left anon

**Comment by barbaldo** [comment]
@marcbouchenoire unfortunately we were never able to isolate the issue. We've figured out with some confidence that the issue does not rely in out code. By that I mean that functions such as `liveblockUsers ` always returns the correct result and only a page reload clears the issue. There's also time involved as in this issue doesn't happen quickly it takes like days of having the browser window open to show up. This is unfortunately the amount of info I'm able to give at the time. If anything else happens I'll post something

**Comment by marcbouchenoire** [comment]
> Maybe there is a way to trigger a clear cache?

We do have [methods](https://liveblocks.io/docs/api-reference/liveblocks-client#Resolvers) to clear all of them or specific keys, maybe @barbaldo you could use them indeed rather than a reload, but that's still a band-aid fix, hopefully we'll find the root problem at some point.

**Comment by barbaldo** [comment]
@marcbouchenoire That's great actually I'll give it a try!

**Comment by onx2** [comment]
We discovered that there is an issue when there's a mismatch of returned # of users and the expected # of users. This happens to us when someone no longer exists in the system but previously added a comment. So to get around that issue we just insert dummy data for the # of missing users -- essentially explicitly saying this person is a deleted user so liveblocks gets the expected # of users back from the endpoint.

**Comment by adrienhongcs** [comment]
Is there a way for Liveblocks to only displays users with missing user info as _Anonymous_?

Currently, one missing user info returned by the **`resolveUsers`** handler leads to all users in a thread displayed as _Anonymous_.

@marcbouchenoire

**Comment by marcbouchenoire** [comment]
@adrienhongcs `resolveUsers` expects the same number of results as inputs and the same order so removing one breaks this, instead you can return `undefined` for user IDs you want to keep as "Anonymous".

```ts
// userIds: ["a", "b", "c"]
resolveUsers={async ({ userIds }) => {
    // ...

    // Only "b" is missing
    return [userA, undefined, userC]
}}
```
